common/hashing.h

```diff
@@ -185,13 +183,30 @@ inline auto HashValue(const T& value) -> HashCode;
 //
 // Example usage:
 // ```cpp
-// auto CarbonHash(const MyType& value, uint64_t seed) -> HashCode {
+// auto CarbonHashValue(const MyType& value, uint64_t seed) -> HashCode {
 //   Hasher hasher(seed);
 //   hasher.HashTwo(value.x, value.y);
 //   return static_cast<HashCode>(hasher);
 // }
 // ```
 //
+// The extension point needs to ensure that values that compare equal (including
+// any comparisons with different types that might be used with a hash table of
+// `YourType` keys) produce the same `HashCode` values.
+//
+// For any two input values that compare unequal, the extension point should
+// maximize the probability of each bit of their resulting `HashCode`s
+// differing. More formally, `HashCode`s should exhibit an [avalanche
+// effect][4]. However, while this is desirable, it should be **secondary** to
```

Comment:
My understanding is that an avalanche effect is about maximizing the number of bits that differ _for inputs that are close together_. Without that qualification, this criterion seems equivalent to requiring that no bit be systematically biased toward 0 or 1, which is a much weaker requirement (for example, the identity function on `int64_t` satisfies it). Is that intentional?

Reply:
This doesn't seem quite right to me, but this may well be a case where I'm getting hung up on technicalities (and/or I'm just mistaken).

> But I also agree with the less formal and less strong guidance of no bit should be systematically biased towards 0 or 1 **for any two inputs that compare unequal**. I don't think that the identify function holds there, because large classes of unequal inputs (small values) have their high bits systematically bias towards 0.

Sure, but that's a different criterion than the one stated in the comment, which doesn't talk about classes of inputs. It's also an unsatisfiable criterion: _any_ possible hash function will have the property that "a large class of unequal inputs have their high bits systematically biased towards 0" -- for example, consider the class of inputs whose hash codes happen to have a high byte equal to 0. For that criterion to be useful, you have to restrict what kinds of classes of inputs can be considered. 

If I understand correctly, the avalanche criterion effectively does that by restricting consideration to classes that are "small", i.e. that consist of inputs that are close to each other in edit distance. As a result, I don't think your rephrased version quite works, because you seem to be saying that all possible classes of inputs can be considered, _including_ classes that are small, but it's crucial that the avalanche criterion _only_ considers classes that are small. 

> So, what change would help clarify here? I'm not sure based on your comment.

See my suggested edit above.